esp32/heltec_wifi_lora_32: Add HELTEC ESP32 WiFi LoRa 32

adalava opened the pull request:
Product page: https://heltec.org/project/wifi-lora-32/

Supported:
    - LED
    - PRG Button
    - OLED Display
 
Unsupported:
    - LoRa

Tested with:
    - samples\hello_world
    - samples\basic\blinky
    - samples\basic\button
    - samples\subsys\display\lvgl

Signed-off-by: Alfredo Dal'Ava Junior <alfredo.junior@eldorado.org.br>


NOTES:
 - Board code is identical to boards/xtensa/esp32
 - Use 'west build -p auto -b heltec_wifi_lora32_v2 samples/subsys/display/lvgl -- -DSHIELD=ssd1306_128x64' do test the OLED display

uLipe commented on boards/xtensa/heltec_wifi_lora32/Kconfig.board:
```diff
@@ -0,0 +1,8 @@
+
+config BOARD_HELTEC_WIFI_LORA32
+	bool "HELTEC WiFi LoRa 32 (V2) Board"
+	depends on SOC_ESP32
```
Missing newline

adalava replied:
done, thanks!

uLipe commented on boards/xtensa/heltec_wifi_lora32/board.cmake:
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: Apache-2.0
+
+include(${ZEPHYR_BASE}/boards/common/esp32.board.cmake)
```
missing newline

uLipe commented on boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32.yaml:
```diff
@@ -0,0 +1,13 @@
+  - watchdog
+  - uart
+  - pinmux
+  - nvs
```
missing newline

uLipe commented on boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32_defconfig:
```diff
@@ -0,0 +1,34 @@
+CONFIG_I2C_1=y
+CONFIG_CLOCK_CONTROL=y
+
+CONFIG_BOOTLOADER_ESP_IDF=y
```
missing newline

uLipe commented:
Some minor comments, otherwise generally LGTM

shubhamkulkarni97 commented on boards/xtensa/heltec_wifi_lora32/CMakeLists.txt:
```diff
@@ -0,0 +1,54 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_BOOTLOADER_ESP_IDF)
```
Is it necessary to duplicate this file? Maybe we can create a common CMake file which can be used by all ESP32 boards.

adalava replied:
@shubhamkulkarni97 you're correct. I copied it from the esp32 directory but I verified there's no problem removing it

shubhamkulkarni97 commented on boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32_defconfig:
```diff
@@ -0,0 +1,34 @@
+CONFIG_XTENSA_RESET_VECTOR=n
+
+CONFIG_BOARD_HELTEC_WIFI_LORA32=y
+CONFIG_SOC_ESP32=y
```
Default configs also seem to be same for all ESP32 boards. We can move it to a common file?

adalava replied:
It's a good idea. I looked at other boards but I couldn't figure out where to put a common file and make both boards/xtensa/esp32 and boards/xtensa/heltec_wifi_lora32_v2 use it. Do you have a recommendation?

adalava replied:
It looks like @sylvioalves is working on some thing like this. Should we keep this PR as is and move the common config on another PR? (The same for CMake files)

sylvioalves replied:
Let's keep as is. Once we refactor board folder structure we re-work this properly.

shubhamkulkarni97 reviewed (commented):
@adalava It will be better to maintain the code if we move ESP32 specific code to a common place. Minor comments, otherwise LGTM.

adalava commented:
Thank you for reviewing it.
I updated the branch to be in sync with main again and added documentation to docs folder, based on the original esp32 port documentation. If something else is needed, just let me know.  I'm also online on slack at #esp32.

sylvioalves commented on boards/xtensa/heltec_wifi_lora32_v2/Kconfig.board:
```diff
@@ -0,0 +1,8 @@
+# HELTEC ESP32 board configuration
+
+# Copyright (c) 2017 Intel Corporation
```
I think you can update this copyright to 2021. Eventually change to the company you work for. If none, ok.

sylvioalves commented on boards/xtensa/heltec_wifi_lora32_v2/doc/index.rst:
```diff
@@ -0,0 +1,160 @@
+.. note::
+
+   In case of manual toolchain installation, set :file:`ESPRESSIF_TOOLCHAIN_PATH` accordingly.
+   Otherwise, set toolchain path as below. If necessary, update the version folder path as in :file:`esp-2020r3-8.4.0`.
```
Can you update the guidelines in this file? Toolchain instructions was changed.

sylvioalves commented on boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts:
```diff
@@ -0,0 +1,154 @@
+	cts-pin = <19>;
+};
+
+&uart1 {
+	current-speed = <115200>;
+	tx-pin = <10>;
+	rx-pin = <9>;
+	rts-pin = <11>;
+	cts-pin = <6>;
+	hw-flow-control;
+};
+
+&uart2 {
+	current-speed = <115200>;
+	tx-pin = <17>;
+	rx-pin = <16>;
+	rts-pin = <7>;
+	cts-pin = <8>;
+	hw-flow-control;
+};
```
From my perspective, you could remove this unused uart instances.

sylvioalves reviewed (dismissed):
Only a few not blocking comments.

adalava commented:
Thanks for the review, @sylvioalves. I like all your points, the diff was updated and rebased.

sylvioalves commented on boards/xtensa/heltec_wifi_lora32_v2/CMakeLists.txt:
```diff
@@ -0,0 +1,65 @@
+
+  board_finalize_runner_args(esp32 "--esp-app-address=0x10000")
+
+endif()
```
Please, delete this file, it was moved into soc folder.

sylvioalves commented on boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts:
```diff
@@ -0,0 +1,86 @@
+	scl-pin = <22>;
+};
+
+&flash0 {
```
You might want to update this if MCUBoot is needed.

sylvioalves commented on boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts:
```diff
@@ -0,0 +1,86 @@
+	current-speed = <115200>;
+	tx-pin = <1>;
+	rx-pin = <3>;
+	rts-pin = <22>;
```
Remove RTS and CTS pin if not used.

sylvioalves commented:
@adalava Also check the commit message, CI is failing due to that.

adalava commented:
> @adalava Also check the commit message, CI is failing due to that.

Thanks! I updated the branch. I hope it's ready now.

sylvioalves commented on boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2_defconfig:
```diff
@@ -0,0 +1,30 @@
+CONFIG_I2C=y
+CONFIG_CLOCK_CONTROL=y
+
+CONFIG_BOOTLOADER_ESP_IDF=y
```
This can be removed as ESP_IDF bootloader is the default.

adalava committed:
esp32/heltec_wifi_lora32_v2: add documentation

Provides basic documentation based on generic board esp32

Signed-off-by: Alfredo Dal'Ava Junior <alfredo.junior@eldorado.org.br>

adalava committed:
esp32/heltec_wifi_lora32_v2: Add HELTEC ESP32 WiFi LoRa 32 (V2)

Product page: https://heltec.org/project/wifi-lora-32/

Supported:
    - LED
    - PRG Button
    - OLED display

Unsupported:
    - LoRa

Tested with:
    - samples\hello_world
    - samples\basic\blinky
    - samples\basic\button
    - samples\boards\esp32\wifi_station
    - samples\subsys\display\lvgl

Signed-off-by: Alfredo Dal'Ava Junior <alfredo.junior@eldorado.org.br>

adalava commented:
Update: Added OLED support out of the box (now possible thanks to @sylvioalves fixes and tips)

@mbolivar-nordic, I described the "vext control" and "oled reset" pins on the DTS under leds/gpio-leds. Do you have any suggestion of better place to describe it? Thanks!

mbolivar-nordic commented on boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts:
```diff
@@ -0,0 +1,126 @@
+		i2c-0 = &i2c0;
+		led0 = &led0;
+		sw0 = &button0;
+		vext = &vext;
```
Do you really need this alias? Why not use `DT_NODELABEL(vext)` to get the node identifier?

adalava replied:
Works with DT_NODELABEL. Thanks!

adalava replied:
> > @mbolivar-nordic, I described the "vext control" and "oled reset" pins on the DTS under leds/gpio-leds. Do you have any suggestion of better place to describe it? Thanks!
> 
> Are they actually LEDs? Or are tehy something else?

They dont describe any LED. The "vext control" controls a mosftet that turns on/off  the power supply for devices external to the microcontroler  (like the onboard OLED display is included).
The "oled reset" pin is the RST pin of the OLED display.

mbolivar-nordic reviewed (commented):
> @mbolivar-nordic, I described the "vext control" and "oled reset" pins on the DTS under leds/gpio-leds. Do you have any suggestion of better place to describe it? Thanks!

Are they actually LEDs? Or are tehy something else?

adalava commented:
Applied @mbolivar-nordic suggestions, rebased and retested the examples.

sylvioalves commented:
@adalava, please, rebase.
